Adding function that creates elipses around clusters for 2 markers

andrewram4287 opened the pull request:
Creating ellipses around clusters of data, including seed to create more consistent intialiazation parametres and fixing poetry to be a lower version of python

aarmey commented:
@andrewram4287 ok, first thing here is have your ellipse function take the center and covariance matrix for the cluster you are plotting. (If it's not clear where to find this, @borcuttjahns did it in his plotting.) You'll want to only keep the two variables you're plotting, so the center is a vector with two elements, and the covariance is a 2-by-2 matrix.

Then you need to convert from covariance matrix to widths and angle. There is a good description of this hidden in [here](https://datascienceplus.com/understanding-the-covariance-matrix/). You want the eigendecomposition of the covariance matrix: `U, S, _ = np.linalg.svd(Cov)`. S[0] and S[1] are the standard deviations along the major and minor axis. The first eigenvector in U points in the direction of the major axis, which you can then convert to an angle.

andrewram4287 commented:
> @andrewram4287 ok, first thing here is have your ellipse function take the center and covariance matrix for the cluster you are plotting. (If it's not clear where to find this, @borcuttjahns did it in his plotting.) You'll want to only keep the two variables you're plotting, so the center is a vector with two elements, and the covariance is a 2-by-2 matrix.
> 
> Then you need to convert from covariance matrix to widths and angle. There is a good description of this hidden in [here](https://datascienceplus.com/understanding-the-covariance-matrix/). You want the eigendecomposition of the covariance matrix: `U, S, _ = np.linalg.svd(Cov)`. S[0] and S[1] are the standard deviations along the major and minor axis. The first eigenvector in U points in the direction of the major axis, which you can then convert to an angle.

I've tried implemented the std so far, but they seem to be a bit off in the plots.

aarmey commented:
> > @andrewram4287 ok, first thing here is have your ellipse function take the center and covariance matrix for the cluster you are plotting. (If it's not clear where to find this, @borcuttjahns did it in his plotting.) You'll want to only keep the two variables you're plotting, so the center is a vector with two elements, and the covariance is a 2-by-2 matrix.
> 
> > 
> 
> > Then you need to convert from covariance matrix to widths and angle. There is a good description of this hidden in [here](https://datascienceplus.com/understanding-the-covariance-matrix/). You want the eigendecomposition of the covariance matrix: `U, S, _ = np.linalg.svd(Cov)`. S[0] and S[1] are the standard deviations along the major and minor axis. The first eigenvector in U points in the direction of the major axis, which you can then convert to an angle.
> 
> 
> 
> I've tried implemented the std so far, but they seem to be a bit off in the plots.

Can you post how so?

andrewram4287 commented:
> > > @andrewram4287 ok, first thing here is have your ellipse function take the center and covariance matrix for the cluster you are plotting. (If it's not clear where to find this, @borcuttjahns did it in his plotting.) You'll want to only keep the two variables you're plotting, so the center is a vector with two elements, and the covariance is a 2-by-2 matrix.
> > 
> > 
> > > 
> > 
> > 
> > > Then you need to convert from covariance matrix to widths and angle. There is a good description of this hidden in [here](https://datascienceplus.com/understanding-the-covariance-matrix/). You want the eigendecomposition of the covariance matrix: `U, S, _ = np.linalg.svd(Cov)`. S[0] and S[1] are the standard deviations along the major and minor axis. The first eigenvector in U points in the direction of the major axis, which you can then convert to an angle.
> > 
> > 
> > I've tried implemented the std so far, but they seem to be a bit off in the plots.
> 
> Can you post how so?

I just committed a version that uses the factors to obtain the means and covariance! The ellipses seem to be centered but, the width/height once again do not seem to be capturing all the data.

aarmey commented:
@andrewram4287 I fixed an issue and cleaned up some. I told you the wrong thing about reducing the covariance matrix; you want to take the projection on both the right and left side.

How do I build the beach scene with ellipses?

andrewram4287 commented:
> @andrewram4287 I fixed an issue and cleaned up some. I told you the wrong thing about reducing the covariance matrix; you want to take the projection on both the right and left side.
> 
> How do I build the beach scene with ellipses?

Just allowed ellipses for the beach scene to occur! But I think the ellipses seem to be a bit off still.

andrewram4287 commented:
@aarmey The U matrix is very small, so the angles are very small and do not move much.

aarmey commented on gmm/figures/common.py:
```diff
@@ -103,15 +103,19 @@ def add_elipse(timei, dosei, ligandi, preNormCP, optPT, marker1, marker2, n_clus
         S = np.sqrt(S)
-
+        
+        angle = np.angle(U[0,0] + complex(U[0,1]),deg=True) 
```
@andrewram4287  (1) This doesn't use complex correctly. (2) Check your indexing for U. You want both indices from the first eigenvector.

aarmey reviewed (approved):
Looks good. Merging.
